lib/services/platform-service.ts

```diff
@@ -313,55 +311,191 @@ export class PlatformService implements IPlatformService {
 		return (() => {
-			this.ensurePlatformInstalled(platform).wait();
-
-			const appSourceDirectoryPath = path.join(this.$projectData.projectDir, constants.APP_FOLDER_NAME);
+			if (this.$options.release || this.$projectChangesService.currentChanges.changesRequireBuild) {
+				return true;
+			}
 			let platformData = this.$platformsData.getPlatformData(platform);
-			let appDestinationDirectoryPath = path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME);
-			const appUpdater = new AppFilesUpdater(appSourceDirectoryPath, appDestinationDirectoryPath, this.$options, this.$fs);
-			appUpdater.cleanDestinationApp();
-		}).future<void>()();
+			let forDevice = !buildConfig || buildConfig.buildForDevice;
+			let outputPath =  forDevice ? platformData.deviceBuildOutputPath : platformData.emulatorBuildOutputPath;
+			if (!this.$fs.exists(outputPath)) {
+				return true;
+			}
+			let packageNames = forDevice ? platformData.validPackageNamesForDevice : platformData.validPackageNamesForEmulator;
+			let packages = this.getApplicationPackages(outputPath, packageNames);
+			if (packages.length === 0) {
+				return true;
+			}
+			let prepareInfo = this.$projectChangesService.getPrepareInfo(platform);
+			let buildInfo = this.getBuildInfo(platform, platformData, buildConfig);
+			if (!prepareInfo || !buildInfo) {
+				return true;
+			}
+			if (this.$options.clean) {
+				return prepareInfo.time !== buildInfo.prepareTime;
+			}
+			if (prepareInfo.time === buildInfo.prepareTime) {
+				return false;
+			}
+			return prepareInfo.changesRequireBuildTime !== buildInfo.prepareTime;
+		}).future<boolean>()();
 	}
 
-	public buildPlatform(platform: string, buildConfig?: IBuildConfig, forceBuild?: boolean): IFuture<void> {
+	public buildPlatform(platform: string, buildConfig?: IBuildConfig): IFuture<void> {
 		return (() => {
-			let shouldBuild = this.preparePlatform(platform, false).wait();
+			this.$logger.out("Building project...");
 			let platformData = this.$platformsData.getPlatformData(platform);
+			platformData.platformProjectService.buildProject(platformData.projectRoot, buildConfig).wait();
+			let prepareInfo = this.$projectChangesService.getPrepareInfo(platform);
 			let buildInfoFilePath = this.getBuildOutputPath(platform, platformData, buildConfig);
 			let buildInfoFile = path.join(buildInfoFilePath, buildInfoFileName);
-			if (!shouldBuild) {
-				if (this.$fs.exists(buildInfoFile)) {
-					let buildInfoText = this.$fs.readText(buildInfoFile);
-					shouldBuild = this._prepareInfo.time !== buildInfoText;
-				} else {
-					shouldBuild = true;
-				}
+			let buildInfo: IBuildInfo = {
+				prepareTime: prepareInfo.changesRequireBuildTime,
+				buildTime: new Date().toString()
+			};
+			this.$fs.writeJson(buildInfoFile, buildInfo);
+			this.$logger.out("Project successfully built.");
+		}).future<void>()();
+	}
+
+	public shouldInstall(device: Mobile.IDevice): boolean {
+		if(this.$options.release) {
+			return true;
```

Comment:
Humm, just wanted to ensure that everything is ok in release. In case if there are some artefacts from debug. It shouldn't be necessary.